src/tmt300rs/hid-tmt300rs.c

```diff
@@ -760,64 +617,54 @@ static int t300rs_update_ramp(struct t300rs_device_entry *t300rs,
 	struct ff_ramp_effect ramp_old = old.u.ramp;
 	struct __packed t300rs_packet_mod_ramp {
 		struct t300rs_packet_header header;
-		uint8_t attribute;
-		uint16_t difference;
-		uint16_t level;
+		uint8_t type;
+		uint16_t slope;
+		uint16_t center;
+		uint16_t length;
+		struct t300rs_packet_envelope envelope;
+		uint8_t effect_type;
+		uint8_t update_type;
+		uint16_t length2;
+		uint16_t offset;
 	} *packet_mod_ramp = (struct t300rs_packet_mod_ramp *)t300rs->send_buffer;
 
-	int ret;
-
-	uint16_t difference, top, bottom;
-	int16_t level;
+	int ret = 0;
 
-	top = ramp.end_level > ramp.start_level ? ramp.end_level : ramp.start_level;
-	bottom = ramp.end_level > ramp.start_level ? ramp.start_level : ramp.end_level;
+	uint8_t invert, old_invert;
+	uint16_t slope, old_slope, length, old_length;
+	int16_t center, old_center;
 
+	t300rs_calculate_ramp_parameters(&slope, &center, &invert, &effect);
+	t300rs_calculate_ramp_parameters(&old_slope, &old_center, &old_invert, &old);
 
-	difference = ((top - bottom) * fixp_sin16(effect.direction * 360 / 0x10000)) / 0x7fff;
+	length = effect.replay.length - 1;
+	old_length = old.replay.length - 1;
 
+	if (slope != old_slope
+			|| center != old_center
+			|| invert != old_invert
+			|| length != old_length
+			|| effect.replay.delay != old.replay.delay
+			|| t300rs_is_envelope_changed(&ramp.envelope, &ramp_old.envelope)) {
```

Comment:
Same nit as in `t300rs_update_constant()`, make `if` do an early return and move `t300rs_is_envelope_changed()` to be checked first.